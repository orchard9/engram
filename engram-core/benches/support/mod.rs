pub mod ann_common;
pub mod datasets;
pub mod engram_ann;

// Real ANN library implementations (require feature flag and system libraries)
#[cfg(feature = "ann_benchmarks")]
pub mod annoy_ann;
#[cfg(feature = "ann_benchmarks")]
pub mod faiss_ann; // Note: Uses mock impl due to annoy-rs API limitations
