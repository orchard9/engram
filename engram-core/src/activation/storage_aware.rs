use crate::activation::ActivationRecord;
use atomic_float::AtomicF32;
use std::convert::TryFrom;
use std::fmt;
use std::sync::atomic::{AtomicU8, AtomicU16, AtomicU64, Ordering};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

/// Storage tier classification for activation spreading
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum StorageTier {
    /// In-memory hot tier with minimal latency.
    Hot = 0,
    /// SSD-backed warm tier balancing capacity and latency.
    Warm = 1,
    /// Columnar cold tier optimized for capacity over latency.
    Cold = 2,
}

impl StorageTier {
    /// Activation threshold that must be exceeded for the tier.
    #[must_use]
    pub const fn activation_threshold(self) -> f32 {
        match self {
            Self::Hot => 0.01,
            Self::Warm => 0.05,
            Self::Cold => 0.1,
        }
    }

    /// Latency budget enforced for the tier.
    #[must_use]
    pub const fn latency_budget(self) -> Duration {
        match self {
            Self::Hot => Duration::from_micros(100),
            Self::Warm => Duration::from_millis(1),
            Self::Cold => Duration::from_millis(10),
        }
    }

    /// Confidence adjustment factor per tier.
    #[must_use]
    pub const fn confidence_factor(self) -> f32 {
        match self {
            Self::Hot => 1.0,
            Self::Warm => 0.95,
            Self::Cold => 0.9,
        }
    }

    /// Derive a tier heuristic based on activation depth.
    #[must_use]
    pub const fn from_depth(depth: u16) -> Self {
        if depth == 0 {
            Self::Hot
        } else if depth <= 2 {
            Self::Warm
        } else {
            Self::Cold
        }
    }
}

/// Atomic bitflag container for activation metadata.
#[derive(Debug, Default)]
pub struct ActivationFlags {
    bits: AtomicU8,
}

impl ActivationFlags {
    /// Flag indicating migration is in progress.
    pub const MIGRATION_PENDING: u8 = 0b0000_0001;
    /// Flag indicating the activation exceeded latency budget.
    pub const LATENCY_OVER_BUDGET: u8 = 0b0000_0010;
    /// Flag indicating the activation attained high confidence.
    pub const HIGH_CONFIDENCE: u8 = 0b0000_0100;
    /// Flag indicating activation was throttled due to tier limits.
    pub const THROTTLED: u8 = 0b0000_1000;

    /// Create a new empty flag container.
    #[must_use]
    pub const fn new() -> Self {
        Self {
            bits: AtomicU8::new(0),
        }
    }

    /// Insert the provided flag bit.
    pub fn insert(&self, flag: u8) {
        let mut current = self.bits.load(Ordering::Relaxed);
        loop {
            let updated = current | flag;
            match self.bits.compare_exchange_weak(
                current,
                updated,
                Ordering::Relaxed,
                Ordering::Relaxed,
            ) {
                Ok(_) => return,
                Err(actual) => current = actual,
            }
        }
    }

    /// Remove the provided flag bit.
    pub fn remove(&self, flag: u8) {
        let mut current = self.bits.load(Ordering::Relaxed);
        let mask = !flag;
        loop {
            let updated = current & mask;
            match self.bits.compare_exchange_weak(
                current,
                updated,
                Ordering::Relaxed,
                Ordering::Relaxed,
            ) {
                Ok(_) => return,
                Err(actual) => current = actual,
            }
        }
    }

    /// Check if the container currently has the flag set.
    #[must_use]
    pub fn contains(&self, flag: u8) -> bool {
        self.bits.load(Ordering::Relaxed) & flag == flag
    }

    /// Clear all stored flags.
    pub fn clear(&self) {
        self.bits.store(0, Ordering::Relaxed);
    }

    /// Return the underlying bit representation.
    #[must_use]
    pub fn bits(&self) -> u8 {
        self.bits.load(Ordering::Relaxed)
    }
}

impl Clone for ActivationFlags {
    fn clone(&self) -> Self {
        Self {
            bits: AtomicU8::new(self.bits.load(Ordering::Relaxed)),
        }
    }
}

impl fmt::Display for ActivationFlags {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:#010b}", self.bits())
    }
}

/// Storage-aware activation record aligned to a cache line.
#[repr(C, align(64))]
#[derive(Debug)]
pub struct StorageAwareActivation {
    /// Identifier for the memory node.
    pub memory_id: String,
    /// Activation level observed for this node.
    pub activation_level: AtomicF32,
    /// Confidence associated with the activation.
    pub confidence: AtomicF32,
    /// Hop count tracking traversal depth.
    pub hop_count: AtomicU16,
    /// Storage tier classification for the memory.
    pub storage_tier: StorageTier,
    /// Metadata flags for migration and throttling status.
    pub flags: ActivationFlags,
    /// Creation instant used for latency calculations.
    pub creation_time: Instant,
    /// Last update timestamp represented as nanoseconds since UNIX epoch.
    pub last_update: AtomicU64,
    /// Observed access latency for the activation.
    pub access_latency: Duration,
    /// Cached confidence multiplier for the tier.
    pub tier_confidence_factor: f32,
}

impl StorageAwareActivation {
    /// Construct a new storage-aware activation for the provided tier.
    #[must_use]
    pub fn new(memory_id: String, storage_tier: StorageTier) -> Self {
        let now = Instant::now();
        let timestamp = current_time_nanos();
        Self {
            memory_id,
            activation_level: AtomicF32::new(0.0),
            confidence: AtomicF32::new(0.0),
            hop_count: AtomicU16::new(0),
            storage_tier,
            flags: ActivationFlags::new(),
            creation_time: now,
            last_update: AtomicU64::new(timestamp),
            access_latency: Duration::default(),
            tier_confidence_factor: storage_tier.confidence_factor(),
        }
    }

    /// Construct activation metadata from a core activation record.
    #[must_use]
    pub fn from_activation_record(record: &ActivationRecord, tier: StorageTier) -> Self {
        let activation = Self::new(record.node_id.clone(), tier);
        let activation_value = record.activation_atomic().load(Ordering::Relaxed);
        activation
            .activation_level
            .store(activation_value, Ordering::Relaxed);

        let confidence_value = record.get_confidence();
        let adjusted_confidence = activation.adjust_confidence_for_tier(confidence_value);
        activation
            .confidence
            .store(adjusted_confidence, Ordering::Relaxed);

        let visits = record.visits_atomic().load(Ordering::Relaxed);
        let hop_value = u16::try_from(visits).unwrap_or(u16::MAX);
        activation.hop_count.store(hop_value, Ordering::Relaxed);

        let timestamp = record.timestamp.load(Ordering::Relaxed);
        if timestamp != 0 {
            activation.last_update.store(timestamp, Ordering::Relaxed);
        }

        activation
    }

    /// Activation threshold derived from the tier configuration.
    #[must_use]
    pub const fn tier_threshold(&self) -> f32 {
        self.storage_tier.activation_threshold()
    }

    /// Adjust the provided confidence score according to tier factor.
    #[must_use]
    pub const fn adjust_confidence_for_tier(&self, confidence: f32) -> f32 {
        (confidence * self.tier_confidence_factor).clamp(0.0, 1.0)
    }

    /// Update activation atomically and refresh metadata timestamps.
    pub fn accumulate_activation(&self, delta: f32) -> f32 {
        let mut current = self.activation_level.load(Ordering::Relaxed);
        loop {
            let updated = (current + delta).clamp(0.0, 1.0);
            if let Err(next) = self.activation_level.compare_exchange_weak(
                current,
                updated,
                Ordering::Relaxed,
                Ordering::Relaxed,
            ) {
                current = next;
                continue;
            }
            self.touch();
            return updated;
        }
    }

    /// Record the observed latency for the activation.
    pub const fn record_latency(&mut self, latency: Duration) {
        self.access_latency = latency;
    }

    /// Increment hop count and return the updated value.
    pub fn increment_hop(&self) -> u16 {
        self.hop_count.fetch_add(1, Ordering::Relaxed) + 1
    }

    /// Determine if the activation fits within the provided latency budget.
    #[must_use]
    pub fn can_access_within_budget(&self, latency_budget: Duration) -> bool {
        self.access_latency <= latency_budget
    }

    /// Update last update timestamp with the current time.
    pub fn touch(&self) {
        self.last_update
            .store(current_time_nanos(), Ordering::Relaxed);
    }

    /// Retrieve the last update timestamp in nanoseconds.
    #[must_use]
    pub fn last_update_nanos(&self) -> u64 {
        self.last_update.load(Ordering::Relaxed)
    }
}

fn current_time_nanos() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .ok()
        .and_then(|duration| u64::try_from(duration.as_nanos()).ok())
        .unwrap_or(u64::MAX)
}

#[cfg(test)]
mod tests {
    use super::{ActivationFlags, StorageAwareActivation, StorageTier};

    #[test]
    fn test_storage_aware_activation_creation() {
        let activation = StorageAwareActivation::new("mem-1".to_string(), StorageTier::Hot);
        assert_eq!(activation.memory_id, "mem-1");
        assert_eq!(activation.storage_tier, StorageTier::Hot);
        assert!(
            activation
                .activation_level
                .load(std::sync::atomic::Ordering::Relaxed)
                .abs()
                < f32::EPSILON
        );
        assert!(
            activation
                .confidence
                .load(std::sync::atomic::Ordering::Relaxed)
                .abs()
                < f32::EPSILON
        );
        assert_eq!(
            activation
                .hop_count
                .load(std::sync::atomic::Ordering::Relaxed),
            0
        );
        assert!((activation.tier_confidence_factor - 1.0).abs() < f32::EPSILON);
    }

    #[test]
    fn test_tier_threshold_values() {
        assert!((StorageTier::Hot.activation_threshold() - 0.01).abs() < f32::EPSILON);
        assert!((StorageTier::Warm.activation_threshold() - 0.05).abs() < f32::EPSILON);
        assert!((StorageTier::Cold.activation_threshold() - 0.1).abs() < f32::EPSILON);
    }

    #[test]
    fn test_confidence_adjustment_per_tier() {
        let hot = StorageAwareActivation::new("h".into(), StorageTier::Hot);
        let warm = StorageAwareActivation::new("w".into(), StorageTier::Warm);
        let cold = StorageAwareActivation::new("c".into(), StorageTier::Cold);

        assert!((hot.adjust_confidence_for_tier(0.8) - 0.8).abs() < f32::EPSILON);
        assert!((warm.adjust_confidence_for_tier(0.8) - 0.76).abs() < 1e-6);
        assert!((cold.adjust_confidence_for_tier(0.8) - 0.72).abs() < 1e-6);
    }

    #[test]
    fn test_atomic_activation_updates() {
        use std::sync::Arc;
        use std::thread;

        let activation = Arc::new(StorageAwareActivation::new("mem".into(), StorageTier::Hot));
        let mut handles = Vec::new();

        for _ in 0..8 {
            let cloned = Arc::clone(&activation);
            handles.push(thread::spawn(move || {
                for _ in 0..1_000 {
                    cloned.accumulate_activation(0.0005);
                }
            }));
        }

        for handle in handles {
            if let Err(panic) = handle.join() {
                std::panic::resume_unwind(panic);
            }
        }

        let value = activation
            .activation_level
            .load(std::sync::atomic::Ordering::Relaxed);
        assert!(value <= 1.0);
        assert!(value > 0.0);
    }

    #[test]
    fn test_activation_flags() {
        let flags = ActivationFlags::new();
        assert_eq!(flags.bits(), 0);
        flags.insert(ActivationFlags::MIGRATION_PENDING);
        assert!(flags.contains(ActivationFlags::MIGRATION_PENDING));
        flags.remove(ActivationFlags::MIGRATION_PENDING);
        assert!(!flags.contains(ActivationFlags::MIGRATION_PENDING));
        flags.insert(ActivationFlags::LATENCY_OVER_BUDGET | ActivationFlags::THROTTLED);
        assert!(flags.contains(ActivationFlags::LATENCY_OVER_BUDGET));
        assert!(flags.contains(ActivationFlags::THROTTLED));
        flags.clear();
        assert_eq!(flags.bits(), 0);
    }

    #[test]
    fn test_tier_from_depth() {
        assert_eq!(StorageTier::from_depth(0), StorageTier::Hot);
        assert_eq!(StorageTier::from_depth(1), StorageTier::Warm);
        assert_eq!(StorageTier::from_depth(2), StorageTier::Warm);
        assert_eq!(StorageTier::from_depth(3), StorageTier::Cold);
        assert_eq!(StorageTier::from_depth(10), StorageTier::Cold);
    }
}
