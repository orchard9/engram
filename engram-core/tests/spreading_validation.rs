#![allow(missing_docs)]
mod support;

use engram_core::activation::storage_aware::StorageTier;
use engram_core::activation::test_support::{
    SpreadingSnapshot, deterministic_config, run_spreading_snapshot,
};
use insta::assert_yaml_snapshot;
use serde::Serialize;
use support::graph_builders::{canonical_fixtures, chain, cycle_with_breakpoint, directed_cycle};

#[derive(Serialize)]
struct FixtureSnapshot {
    metadata: support::graph_builders::GraphFixtureMetadata,
    snapshot: NormalizedSnapshot,
}

#[derive(Serialize, Clone)]
struct NormalizedSnapshot {
    seeds: Vec<(String, f32)>,
    activations: Vec<NormalizedActivation>,
    tier_summaries: Vec<(String, NormalizedTierSummary)>,
    cycle_paths: Vec<Vec<String>>,
    deterministic_trace: Vec<NormalizedTraceEntry>,
    metrics: NormalizedMetrics,
}

#[derive(Debug, Serialize, Clone, PartialEq)]
struct NormalizedActivation {
    memory_id: String,
    activation: f32,
    confidence: f32,
    hop_count: u16,
    storage_tier: String,
    flags: String,
}

#[derive(Serialize, Clone)]
struct NormalizedTierSummary {
    node_count: usize,
    total_activation: f32,
    average_confidence: f32,
}

#[derive(Serialize, Clone)]
struct NormalizedTraceEntry {
    depth: u16,
    target_node: String,
    activation: f32,
    confidence: f32,
    source_node: Option<String>,
}

#[derive(Serialize, Clone)]
struct NormalizedMetrics {
    total_activations: u64,
    cache_hits: u64,
    cache_misses: u64,
    average_latency_ns: u64,
    latency_budget_violations: u64,
    cycles_detected: u64,
    parallel_efficiency: f32,
    pool_available: u64,
    pool_in_flight: u64,
    pool_high_water_mark: u64,
}

fn normalize(snapshot: &SpreadingSnapshot) -> NormalizedSnapshot {
    let mut activations: Vec<NormalizedActivation> = snapshot
        .activations
        .iter()
        .map(|node| NormalizedActivation {
            memory_id: node.memory_id.clone(),
            activation: (node.activation * 1_000.0).round() / 1_000.0,
            confidence: (node.confidence * 1_000.0).round() / 1_000.0,
            hop_count: node.hop_count,
            storage_tier: node.storage_tier.clone(),
            flags: node.flags.clone(),
        })
        .collect();
    activations.sort_by(|a, b| a.memory_id.cmp(&b.memory_id));

    let mut tier_summaries: Vec<(String, NormalizedTierSummary)> = snapshot
        .tier_summaries
        .iter()
        .map(|(tier, summary)| {
            (
                tier.clone(),
                NormalizedTierSummary {
                    node_count: summary.node_count,
                    total_activation: (summary.total_activation * 1_000.0).round() / 1_000.0,
                    average_confidence: (summary.average_confidence * 1_000.0).round() / 1_000.0,
                },
            )
        })
        .collect();
    tier_summaries.sort_by(|a, b| a.0.cmp(&b.0));

    let mut deterministic_trace: Vec<NormalizedTraceEntry> = snapshot
        .deterministic_trace
        .iter()
        .map(|entry| NormalizedTraceEntry {
            depth: entry.depth,
            target_node: entry.target_node.clone(),
            activation: (entry.activation * 1_000.0).round() / 1_000.0,
            confidence: (entry.confidence * 1_000.0).round() / 1_000.0,
            source_node: entry.source_node.clone(),
        })
        .collect();
    deterministic_trace.sort_by(|a, b| {
        a.depth
            .cmp(&b.depth)
            .then(a.target_node.cmp(&b.target_node))
    });

    let mut seeds = snapshot.seeds.clone();
    seeds.sort_by(|a, b| a.0.cmp(&b.0));

    let mut cycle_paths = snapshot.cycle_paths.clone();
    cycle_paths.sort();

    NormalizedSnapshot {
        seeds,
        activations,
        tier_summaries,
        cycle_paths,
        deterministic_trace,
        metrics: NormalizedMetrics {
            total_activations: snapshot.metrics.total_activations,
            cache_hits: snapshot.metrics.cache_hits,
            cache_misses: snapshot.metrics.cache_misses,
            average_latency_ns: 0,
            latency_budget_violations: snapshot.metrics.latency_budget_violations,
            cycles_detected: snapshot.metrics.cycles_detected,
            parallel_efficiency: (snapshot.metrics.parallel_efficiency * 1_000.0).round() / 1_000.0,
            pool_available: snapshot.metrics.pool_available,
            pool_in_flight: snapshot.metrics.pool_in_flight,
            pool_high_water_mark: snapshot.metrics.pool_high_water_mark,
        },
    }
}

#[test]
#[ignore = "Flaky: Snapshots change between runs due to non-deterministic spreading depth. Root cause: Same parallel spreading regression as deterministic_chain_runs_consistently - sometimes spreads to 2 nodes, sometimes 4. Needs fix in engram-core/src/activation/parallel.rs"]
fn canonical_spreading_snapshots_are_stable() {
    let base_config = deterministic_config(4242);
    let mut base_config = base_config;
    base_config.num_threads = 1;
    base_config.batch_size = 1;
    base_config.work_stealing_ratio = 0.0;
    base_config.max_concurrent_per_tier = 1;

    for fixture in canonical_fixtures() {
        let mut config = base_config.clone();
        fixture.apply_config_adjustments(&mut config);
        let snapshot = run_spreading_snapshot(fixture.graph(), &fixture.seeds, config)
            .expect("spreading should succeed");
        let metadata = fixture.metadata();
        let normalized = normalize(&snapshot);

        insta::with_settings!({
            snapshot_path => "data/spreading_snapshots",
            snapshot_suffix => "yaml",
            prepend_module_to_snapshot => false,
        }, {
            assert_yaml_snapshot!(
                format!("{}_snapshot", metadata.name),
                FixtureSnapshot {
                    metadata: metadata.clone(),
                    snapshot: normalized,
                }
            );
        });
    }
}

#[test]
#[ignore = "Flaky: Sometimes returns 4 activations, sometimes 2. Root cause: Non-deterministic behavior in parallel spreading engine even with deterministic config. Regression from recent parallel changes. Needs investigation in engram-core/src/activation/parallel.rs"]
fn deterministic_chain_runs_consistently() {
    let fixture = chain(4);
    let mut config = deterministic_config(7);
    fixture.apply_config_adjustments(&mut config);

    let snapshot_a = run_spreading_snapshot(fixture.graph(), &fixture.seeds, config.clone())
        .expect("spreading should succeed");
    let snapshot_b = run_spreading_snapshot(fixture.graph(), &fixture.seeds, config)
        .expect("spreading should succeed");

    assert_eq!(
        normalize(&snapshot_a).activations,
        normalize(&snapshot_b).activations
    );
    assert_eq!(snapshot_a.cycle_paths, snapshot_b.cycle_paths);
    assert_eq!(
        snapshot_a.deterministic_trace.len(),
        snapshot_b.deterministic_trace.len()
    );
}

#[test]
#[ignore = "Known issue: Cycle detection not populating cycle_paths in snapshots. Root cause: CycleDetector.get_cycle_paths() returns empty even when cycles detected. Needs investigation in engram-core/src/activation/cycle_detector.rs"]
fn cycle_breakpoints_surface_cycle_paths() {
    let fixture = cycle_with_breakpoint(6);
    let mut config = deterministic_config(99);
    fixture.apply_config_adjustments(&mut config);

    let snapshot =
        run_spreading_snapshot(fixture.graph(), &fixture.seeds, config).expect("cycle run");

    assert!(
        !snapshot.cycle_paths.is_empty(),
        "expected breakpoint cycle metadata to be captured"
    );
}

#[test]
#[ignore = "Flaky: Timeout waiting for spreading completion (270s). Root cause: Same parallel spreading regression affecting all other tests. Threading error in engram-core/src/activation/parallel.rs wait_for_completion(). Run with: cargo test --test spreading_validation tier_summaries_capture_hot_tier_activity -- --ignored --nocapture"]
fn tier_summaries_capture_hot_tier_activity() {
    let fixture = directed_cycle(5);
    let mut config = deterministic_config(1337);
    fixture.apply_config_adjustments(&mut config);

    let snapshot =
        run_spreading_snapshot(fixture.graph(), &fixture.seeds, config).expect("cycle run");

    let hot_summary = snapshot
        .tier_summaries
        .get(&format!("{:?}", StorageTier::Hot))
        .expect("hot tier summary present");
    assert!(hot_summary.node_count >= 1);
    assert!(snapshot.metrics.total_activations >= 1);
}
