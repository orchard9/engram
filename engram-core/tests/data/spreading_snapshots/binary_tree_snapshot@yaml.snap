---
source: engram-core/tests/spreading_validation.rs
expression: "FixtureSnapshot { metadata: metadata.clone(), snapshot: normalized, }"
---
metadata:
  name: binary_tree
  description: Balanced binary tree fixture for tier coverage
  seeds:
    - - TreeRoot
      - 1
  node_count: 15
  edge_count: 14
  average_out_degree: 0.93333334
  adjacency:
    TreeRoot:
      - target: TreeRoot_left_1
        weight: 0.8
        edge_type: Excitatory
      - target: TreeRoot_right_1
        weight: 0.8
        edge_type: Excitatory
    TreeRoot_left_1:
      - target: TreeRoot_left_1_left_2
        weight: 0.8
        edge_type: Excitatory
      - target: TreeRoot_left_1_right_2
        weight: 0.8
        edge_type: Excitatory
    TreeRoot_left_1_left_2:
      - target: TreeRoot_left_1_left_2_left_3
        weight: 0.8
        edge_type: Excitatory
      - target: TreeRoot_left_1_left_2_right_3
        weight: 0.8
        edge_type: Excitatory
    TreeRoot_left_1_left_2_left_3: []
    TreeRoot_left_1_left_2_right_3: []
    TreeRoot_left_1_right_2:
      - target: TreeRoot_left_1_right_2_left_3
        weight: 0.8
        edge_type: Excitatory
      - target: TreeRoot_left_1_right_2_right_3
        weight: 0.8
        edge_type: Excitatory
    TreeRoot_left_1_right_2_left_3: []
    TreeRoot_left_1_right_2_right_3: []
    TreeRoot_right_1:
      - target: TreeRoot_right_1_left_2
        weight: 0.8
        edge_type: Excitatory
      - target: TreeRoot_right_1_right_2
        weight: 0.8
        edge_type: Excitatory
    TreeRoot_right_1_left_2:
      - target: TreeRoot_right_1_left_2_left_3
        weight: 0.8
        edge_type: Excitatory
      - target: TreeRoot_right_1_left_2_right_3
        weight: 0.8
        edge_type: Excitatory
    TreeRoot_right_1_left_2_left_3: []
    TreeRoot_right_1_left_2_right_3: []
    TreeRoot_right_1_right_2:
      - target: TreeRoot_right_1_right_2_left_3
        weight: 0.8
        edge_type: Excitatory
      - target: TreeRoot_right_1_right_2_right_3
        weight: 0.8
        edge_type: Excitatory
    TreeRoot_right_1_right_2_left_3: []
    TreeRoot_right_1_right_2_right_3: []
snapshot:
  seeds:
    - - TreeRoot
      - 1
  activations:
    - memory_id: TreeRoot
      activation: 1
      confidence: 1
      hop_count: 0
      storage_tier: Hot
      flags: 0b00000000
    - memory_id: TreeRoot_left_1
      activation: 0.397
      confidence: 0.377
      hop_count: 1
      storage_tier: Warm
      flags: 0b00000000
    - memory_id: TreeRoot_left_1_left_2
      activation: 0.078
      confidence: 0.074
      hop_count: 1
      storage_tier: Warm
      flags: 0b00000000
    - memory_id: TreeRoot_left_1_left_2_left_3
      activation: 0
      confidence: 0
      hop_count: 0
      storage_tier: Cold
      flags: 0b00000000
    - memory_id: TreeRoot_left_1_left_2_right_3
      activation: 0
      confidence: 0
      hop_count: 0
      storage_tier: Cold
      flags: 0b00000000
    - memory_id: TreeRoot_left_1_right_2
      activation: 0.078
      confidence: 0.074
      hop_count: 1
      storage_tier: Warm
      flags: 0b00000000
    - memory_id: TreeRoot_left_1_right_2_left_3
      activation: 0
      confidence: 0
      hop_count: 0
      storage_tier: Cold
      flags: 0b00000000
    - memory_id: TreeRoot_left_1_right_2_right_3
      activation: 0
      confidence: 0
      hop_count: 0
      storage_tier: Cold
      flags: 0b00000000
    - memory_id: TreeRoot_right_1
      activation: 0.397
      confidence: 0.377
      hop_count: 1
      storage_tier: Warm
      flags: 0b00000000
    - memory_id: TreeRoot_right_1_left_2
      activation: 0.078
      confidence: 0.074
      hop_count: 1
      storage_tier: Warm
      flags: 0b00000000
    - memory_id: TreeRoot_right_1_left_2_left_3
      activation: 0
      confidence: 0
      hop_count: 0
      storage_tier: Cold
      flags: 0b00000000
    - memory_id: TreeRoot_right_1_left_2_right_3
      activation: 0
      confidence: 0
      hop_count: 0
      storage_tier: Cold
      flags: 0b00000000
    - memory_id: TreeRoot_right_1_right_2
      activation: 0.078
      confidence: 0.074
      hop_count: 1
      storage_tier: Warm
      flags: 0b00000000
    - memory_id: TreeRoot_right_1_right_2_left_3
      activation: 0
      confidence: 0
      hop_count: 0
      storage_tier: Cold
      flags: 0b00000000
    - memory_id: TreeRoot_right_1_right_2_right_3
      activation: 0
      confidence: 0
      hop_count: 0
      storage_tier: Cold
      flags: 0b00000000
  tier_summaries:
    - - Cold
      - node_count: 8
        total_activation: 0
        average_confidence: 0
    - - Hot
      - node_count: 1
        total_activation: 1
        average_confidence: 1
    - - Warm
      - node_count: 6
        total_activation: 1.108
        average_confidence: 0.175
  cycle_paths: []
  deterministic_trace:
    - depth: 1
      target_node: TreeRoot_left_1
      activation: 0.397
      confidence: 0.397
      source_node: TreeRoot
    - depth: 1
      target_node: TreeRoot_right_1
      activation: 0.397
      confidence: 0.397
      source_node: TreeRoot
    - depth: 2
      target_node: TreeRoot_left_1_left_2
      activation: 0.078
      confidence: 0.078
      source_node: TreeRoot_left_1
    - depth: 2
      target_node: TreeRoot_left_1_right_2
      activation: 0.078
      confidence: 0.078
      source_node: TreeRoot_left_1
    - depth: 2
      target_node: TreeRoot_right_1_left_2
      activation: 0.078
      confidence: 0.078
      source_node: TreeRoot_right_1
    - depth: 2
      target_node: TreeRoot_right_1_right_2
      activation: 0.078
      confidence: 0.078
      source_node: TreeRoot_right_1
  metrics:
    total_activations: 6
    cache_hits: 0
    cache_misses: 0
    average_latency_ns: 0
    latency_budget_violations: 0
    cycles_detected: 0
    parallel_efficiency: 0
    pool_available: 10000
    pool_in_flight: 0
    pool_high_water_mark: 10000
